File: 2677-cousins-in-binary-tree-ii/2677-cousins-in-binary-tree-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* replaceValueInTree(TreeNode* root) {
        vector<int> levelSum;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            int n = q.size(), sum = 0;
            for(int i=0; i<n; i++){
                TreeNode *node = q.front(); q.pop();
                sum += node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            levelSum.push_back(sum);
        }
        q.push(root);
        int levelIndex = 1;
        root->val = 0;
        while(!q.empty()){
            int levelSize = q.size(), siblingSum = 0;
            for(int i=0; i<levelSize; i++){
                TreeNode *node = q.front(); q.pop();
                siblingSum = (node->left ? node->left->val : 0) + (node->right ? node->right->val : 0);
                if(node->left){
                    node->left->val = levelSum[levelIndex] - siblingSum;
                    q.push(node->left);
                }
                if(node->right){
                    node->right->val = levelSum[levelIndex] - siblingSum;
                    q.push(node->right);
                }
            }
            levelIndex++;
        }
        return root;
    }
};

File: 2677-cousins-in-binary-tree-ii/2677-cousins-in-binary-tree-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode replaceValueInTree(TreeNode root) {
        int prevLevelSum = root.val;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        while(!q.isEmpty()){
            int levelSize = q.size(), currSum = 0;
            for(int i=0; i<levelSize; i++){
                TreeNode node = q.poll();
                node.val = prevLevelSum - node.val;
                int siblingSum = (node.left != null ? node.left.val : 0) + (node.right!=null ? node.right.val : 0);

                if(node.left != null){
                    currSum += node.left.val;
                    node.left.val = siblingSum;
                    q.offer(node.left);
                }
                if(node.right!=null){
                    currSum += node.right.val;
                    node.right.val = siblingSum;
                    q.offer(node.right);
                }
            }
            prevLevelSum = currSum;
        }
        return root;
    }
}